python/spec/templates/jax/xla_ops.cpp

```diff
@@ -1,169 +1,197 @@
+// Generated JAX FFI bindings for celerite2.
+// Regenerate with: python python/spec/generate.py
+
 #include <pybind11/pybind11.h>
-#include <cstdint>
-#include <stdexcept>
+#include <Eigen/Core>
 #include <string>
-#include <type_traits>
+
+#include "xla/ffi/api/ffi.h"
+#include "xla/ffi/api/api.h"
+
 #include "../driver.hpp"
 
 namespace py = pybind11;
+namespace ffi = xla::ffi;
 using namespace celerite2::driver;
 
+// Helpers
+template <typename Buffer>
+inline Eigen::Index dim0(const Buffer& buf) {
+  return static_cast<Eigen::Index>(buf.dimensions()[0]);
+}
+template <typename Buffer>
+inline Eigen::Index dim1(const Buffer& buf) {
+  return static_cast<Eigen::Index>(buf.dimensions()[1]);
+}
+template <typename Buffer>
+inline Eigen::Index flat_cols(const Buffer& buf) {
+  const auto& dims = buf.dimensions();
+  Eigen::Index cols = 1;
+  for (size_t i = 1; i < dims.size(); ++i)
+    cols *= static_cast<Eigen::Index>(dims[i]);
+  return cols;
+}
+
+inline ffi::Error shape_error(const char* msg) {
+  return ffi::Error(ffi::ErrorCode::kInvalidArgument, std::string(msg));
```

Comment:
```suggestion
  return ffi::Error::InvalidArgument(std::string(msg));
```

But I'd probably just inline this directly below instead of using this helper.